storage/transfermanager/downloader.go

```diff
@@ -69,6 +73,82 @@ func (d *Downloader) DownloadObject(ctx context.Context, input *DownloadObjectIn
 	return nil
 }
 
+// DownloadDirectory queues the download of a set of objects to a local path.
+// This will initiate the download but is non-blocking; call Downloader.Results
+// or use the callback to process the result. DownloadDirectory is thread-safe
+// and can be called simultaneously from different goroutines.
+// DownloadDirectory will resolve any filters on the input and create the needed
+// directory structure locally as the operations progress.
+// Note: DownloadDirectory overwrites existing files in the directory.
+func (d *Downloader) DownloadDirectory(ctx context.Context, input *DownloadDirectoryInput) error {
+	if d.config.asynchronous && input.Callback == nil {
+		return errors.New("transfermanager: input.Callback must not be nil when the WithCallbacks option is set")
+	}
+	if !d.config.asynchronous && input.Callback != nil {
+		return errors.New("transfermanager: input.Callback must be nil unless the WithCallbacks option is set")
+	}
+
+	select {
+	case <-d.doneReceivingInputs:
+		return errors.New("transfermanager: WaitAndClose called before DownloadDirectory")
+	default:
+	}
+
+	objectsToQueue := []DownloadObjectInput{}
+	query := &storage.Query{
+		Prefix:      input.Prefix,
+		StartOffset: input.StartOffset,
+		EndOffset:   input.EndOffset,
+		MatchGlob:   input.MatchGlob,
+	}
+	if err := query.SetAttrSelection([]string{"Name"}); err != nil {
+		return fmt.Errorf("transfermanager: DownloadDirectory query.SetAttrSelection: %w", err)
+	}
+
+	it := d.client.Bucket(input.Bucket).Objects(ctx, query)
+
+	// TODO: Clean up any created directory structure on failure.
+
+	attrs, err := it.Next()
+	for err != iterator.Done {
+		if err != nil {
+			return fmt.Errorf("transfermanager: DownloadDirectory failed to list objects: %w", err)
+		}
+
+		// Add generation?
+
+		object := attrs.Name
+
+		// Make sure all directories in the object path exist.
+		objDirectory := filepath.Join(input.LocalDirectory, filepath.Dir(object))
+		err = os.MkdirAll(objDirectory, fs.ModeDir|fs.ModePerm)
+		if err != nil {
+			return fmt.Errorf("transfermanager: DownloadDirectory failed to make directory(%q): %w", objDirectory, err)
+		}
+
+		// Create file to download to.
+		filePath := filepath.Join(input.LocalDirectory, object)
+		f, fErr := os.Create(filePath)
```

Comment:
Does this work to create subdirs if the object name has slashes?

Reply:
The subdirs should be created in L124 `os.MkdirAll`

Reply:
Gotcha, sounds good.